newlib/libc/stdlib/mallocr.c

```diff
@@ -3597,6 +3602,27 @@ int mALLOPt(param_number, value) RDECL int param_number; int value;
 
 #endif /* DEFINE_MALLOPT */
 
+#ifdef DEFINE_POSIX_MEMALIGN
+#if __STD_C
+int pOSIx_mEMALIGn(Void_t **memptr, size_t align, size_t size)
+#else
+int pOSIx_mEMALIGn(memptr, align, size) RDECL Void_t **memptr; size_t align, size;
+#endif
+{
+    /* Return EINVAL if align isn't power of 2 or not a multiple of a pointer size */
+    if ((align & (align-1)) != 0 || align % sizeof(Void_t *) != 0 || align == 0)
+        return EINVAL;
+
+    Void_t *mem = mEMALIGn(align, size);
```

Comment:
posix_memalign should not set errno on failure; memalign does. Looks like this is also a bug in nano-mallocr. Looking at glibc, it *also* has this bug. I think we punt.

Reply:
The implementation I've added here is adapted directly from the existing one in nano-mallocr, so yes, if this has a bug that I didn't notice (sorry!) then I'm not surprised nano-mallocr has the same one.

By "punt", I guess you mean at least not trying to fix this as part of the same commit?

Reply:
Sorry, an americanism slipped into my message. Yes, given that every other implementation I've looked at uses 'malloc' directly, they'll all end up setting errno unless they carefully save and restore it and so there's no urgent need to fix this soon (if ever). Given that the error reporting from posix_memalign is so unusual, I'm not terribly surprised by this...